### ISSUE TITLE
xdg-desktop-portal 1.8.0 crashing (under Sway?)

### ISSUE THREAD

**Issue by colemickens** [NixOS/nixpkgs#101297]
**Describe the bug**

I noticed that Firefox screensharing wasn't working, despite https://github.com/NixOS/nixpkgs/pull/84233 and my system being configured properly.

Later, I noticed that file uploads were also broken (because the file dialog was not presenting).

Finally later I noticed something coredumping a lot. Turns out it was `xdg-desktop-portal`.

(Note, I ALSO had issues with `/run/user/1000/doc` being mounted ahead of time by some other unit, when `xdg-document-portal` was expecting to handle the mount.)

```
           PID: 31232 (.xdg-desktop-po)
           UID: 1000 (cole)
           GID: 1000 (cole)
        Signal: 11 (SEGV)
     Timestamp: Wed 2020-10-21 16:19:24 PDT (9s ago)
  Command Line: /nix/store/9qbfra5ab43h8kxsdcp4vn8lzh0j5vpn-xdg-desktop-portal-1.8.0/libexec/xdg-desktop-portal
    Executable: /nix/store/9qbfra5ab43h8kxsdcp4vn8lzh0j5vpn-xdg-desktop-portal-1.8.0/libexec/.xdg-desktop-portal-wrapped
 Control Group: /user.slice/user-1000.slice/session-1.scope
          Unit: session-1.scope
         Slice: user-1000.slice
       Session: 1
     Owner UID: 1000 (cole)
       Boot ID: 8b425569d10445f89a268e762876be97
    Machine ID: 17796969210f4298a4d3f91179b0ed70
      Hostname: slynux
       Storage: /var/lib/systemd/coredump/core.\x2exdg-desktop-po.1000.8b425569d10445f89a268e762876be97.31232.1603322364000000000000.lz4
       Message: Process 31232 (.xdg-desktop-po) of user 1000 dumped core.
```

```
(gdb) bt
#0  0x0000000000419aab in settings_handle_read_all ()
#1  0x00007f612d71aabd in ffi_call_unix64 () from /nix/store/s0mblhs5vmjza9dmipn74rwqflxy1fw7-libffi-3.3/lib/libffi.so.7
#2  0x00007f612d71979c in ffi_call_int () from /nix/store/s0mblhs5vmjza9dmipn74rwqflxy1fw7-libffi-3.3/lib/libffi.so.7
#3  0x00007f612dc626ec in g_cclosure_marshal_generic () from /nix/store/vlsc0gm44gdvw7l69yk4z3xbwfvk3v13-glib-2.64.5/lib/libgobject-2.0.so.0
#4  0x00007f612dc61ec2 in g_closure_invoke () from /nix/store/vlsc0gm44gdvw7l69yk4z3xbwfvk3v13-glib-2.64.5/lib/libgobject-2.0.so.0
#5  0x00007f612dc7453e in signal_emit_unlocked_R () from /nix/store/vlsc0gm44gdvw7l69yk4z3xbwfvk3v13-glib-2.64.5/lib/libgobject-2.0.so.0
#6  0x00007f612dc7f358 in g_signal_emitv () from /nix/store/vlsc0gm44gdvw7l69yk4z3xbwfvk3v13-glib-2.64.5/lib/libgobject-2.0.so.0
#7  0x000000000045b38c in _xdp_settings_skeleton_handle_method_call ()
#8  0x00007f612ddc83b0 in dispatch_in_thread_func () from /nix/store/vlsc0gm44gdvw7l69yk4z3xbwfvk3v13-glib-2.64.5/lib/libgio-2.0.so.0
#9  0x00007f612dd5c13e in g_task_thread_pool_thread () from /nix/store/vlsc0gm44gdvw7l69yk4z3xbwfvk3v13-glib-2.64.5/lib/libgio-2.0.so.0
#10 0x00007f612dba0c34 in g_thread_pool_thread_proxy () from /nix/store/vlsc0gm44gdvw7l69yk4z3xbwfvk3v13-glib-2.64.5/lib/libglib-2.0.so.0
#11 0x00007f612dba03cd in g_thread_proxy () from /nix/store/vlsc0gm44gdvw7l69yk4z3xbwfvk3v13-glib-2.64.5/lib/libglib-2.0.so.0
#12 0x00007f612db0aead in start_thread () from /nix/store/9df65igwjmf2wbw0gbrrgair6piqjgmi-glibc-2.31/lib/libpthread.so.0
#13 0x00007f612da3bd2f in clone () from /nix/store/9df65igwjmf2wbw0gbrrgair6piqjgmi-glibc-2.31/lib/libc.so.6
```

Unfortunately, I don't know how to do the debug symbol bit as requested by @jtojnar [here](https://github.com/flatpak/xdg-desktop-portal/issues/501#issuecomment-713469287). I should probably know/learn how to do this.

**To Reproduce**

I don't know exact steps. #84233 documents the rough steps to enable pipewire+firefox scenario.

I don't know what broke this. It worked when #84233 was authored/merged and for some time after. **Note** that this problem is independent of `firefox`, and occurs when other applications trigger paths that reach out to xdg portals (presumably via dbus).

**Expected behavior**

XDG Portal not to crash, and XDG Portal GTK/WLR to handle portal requests properly.

**Metadata**
It's current `nixos-unstable-small`. 8294af02845210169290240e6fd356f6b7ea3f11

Maintainer information:
ccing people who've touched the xdg module: cc: @worldofpeace @jtojnar 

(if anyone knows `hexa-` from `freenode`'s GitHub handle, ping them, they were having this issue too I think)

**Comment by jtojnar** [comment]
You can have debug symbols by using something like this:

```
{
  environment.systemPackages = [
    pkgs.xdg-desktop-portal # install debug symbols
  ];
  environment.enableDebugInfo = true;
  nixpkgs.overlays = [
    (final: prev: {
      xdg-desktop-portal = prev.enableDebugging prev.xdg-desktop-portal;
    })
  ];
}
```

**Comment by colemickens** [comment]
I had to logout/login to get the new build. Also, the issue went away. I've removed the debug stuff I added, rebuilt, we'll see if this start repro-ing again...

**Comment by bnjmnt4n** [comment]
Hey again @colemickens! Coming here from https://github.com/NixOS/nixpkgs/pull/84233, wanted to note that I too cannot run Firefox with screen-sharing. In fact, not too sure if this is a similar problem but I don't seem to have a `xdg-desktop-portal` binary in my PATH at all. Wanted to see if there's any progress in your repro?

**Comment by colemickens** [comment]
I don't have `xdg-desktop-portal` in PATH either. Please see the instructions in #84233 that describe how to properly enable the scenario (which runs `xdg-desktop-portal` as a user service), and then maybe open a different issue.

So far this issue doesn't repro, even after removing the debugging stuff. I don't know if it was a point-in-time issue, if I had some outer space rays hit my RAM or what. I'm going to go ahead and close this.

**Comment by amxj9** [comment]
So, I have this same issue every time with xdg-desktop-portal coredumping when I run e.g. firefox or zathura (config as per #84233, GTK_USE_PORTAL=1).  I've enabled the debug symbols as suggested and the coredump backtrace is as follows:
```
(gdb) bt
#0  0x0000000000418bc2 in settings_handle_read_all (object=<optimized out>, invocation=0x11ae440, arg_namespaces=0x7fb0a0002570) at src/settings.c:73
#1  0x00007fb0ba52fabd in ffi_call_unix64 () from /nix/store/gr98m9k9385hjb5mq1dd2cl1b419xzas-libffi-3.3/lib/libffi.so.7
#2  0x00007fb0ba52e79c in ffi_call_int () from /nix/store/gr98m9k9385hjb5mq1dd2cl1b419xzas-libffi-3.3/lib/libffi.so.7
#3  0x00007fb0baa7b6ec in g_cclosure_marshal_generic (closure=0x120a5f0, return_gvalue=0x7fb0a8f48ab0, n_param_values=<optimized out>, param_values=<optimized out>, 
    invocation_hint=<optimized out>, marshal_data=<optimized out>) at ../gobject/gclosure.c:1500
#4  0x00007fb0baa7aec2 in g_closure_invoke (closure=0x120a5f0, return_value=0x7fb0a8f48ab0, n_param_values=3, param_values=0x7fb0a00021a0, invocation_hint=0x7fb0a8f48a90)
    at ../gobject/gclosure.c:810
#5  0x00007fb0baa8d53e in signal_emit_unlocked_R (node=node@entry=0x120a620, detail=detail@entry=0, instance=instance@entry=0x11d5d20, 
    emission_return=emission_return@entry=0x7fb0a8f48be0, instance_and_params=instance_and_params@entry=0x7fb0a00021a0) at ../gobject/gsignal.c:3780
#6  0x00007fb0baa98358 in g_signal_emitv (instance_and_params=instance_and_params@entry=0x7fb0a00021a0, signal_id=signal_id@entry=57, detail=detail@entry=0, 
    return_value=return_value@entry=0x7fb0a8f48be0) at ../gobject/gsignal.c:3230
#7  0x000000000045a55d in _xdp_settings_skeleton_handle_method_call (connection=<optimized out>, sender=sender@entry=0x7fb09c00c220 ":1.23", 
    object_path=object_path@entry=0x7fb09c0114b0 "/org/freedesktop/portal/desktop", interface_name=interface_name@entry=0x7fb09c010e00 "org.freedesktop.portal.Settings", 
    method_name=method_name@entry=0x7fb09c00eae0 "ReadAll", parameters=parameters@entry=0x7fb094009240, invocation=0x11ae440, user_data=0x11d5d20) at src/xdp-dbus.c:39793
#8  0x00007fb0babe13c0 in dispatch_in_thread_func (task=<optimized out>, source_object=<optimized out>, task_data=0x7fb09c001d20, cancellable=<optimized out>)
    at ../gio/gdbusinterfaceskeleton.c:540
#9  0x00007fb0bab7514e in g_task_thread_pool_thread (thread_data=0x11b9490, pool_data=<optimized out>) at ../gio/gtask.c:1412
#10 0x00007fb0ba9b9c44 in g_thread_pool_thread_proxy (data=<optimized out>) at ../glib/gthreadpool.c:354
#11 0x00007fb0ba9b93dd in g_thread_proxy (data=0x120fc00) at ../glib/gthread.c:807
#12 0x00007fb0ba923e9e in start_thread (arg=0x7fb0a8f49640) at pthread_create.c:463
#13 0x00007fb0ba85266f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb) 
```

**Comment by jtojnar** [comment]
The line is weird though, there is only one dereference so it looks like `xdp_impl_settings_call_read_all_sync` returns `false` but does not set an error?

https://github.com/flatpak/xdg-desktop-portal/blob/1.8.0/src/settings.c#L73

**Comment by amxj9** [comment]
Is there's any further debug info I might be able to provide that will help with this?

**Comment by jtojnar** [comment]
There does not really seem to be a place where the call can fail:

https://github.com/flatpak/xdg-desktop-portal-gtk/blob/5e45e780eecacfe199ee9582cce4c54a27ce186f/src/settings.c#L141-L143

I guess the next step would be monitoring the session bus with `dbus-monitor` or `bustle` and seeing what the portal implementation sends across the wire.

**Comment by colemickens** [comment]
maybe related: https://github.com/emersion/xdg-desktop-portal-wlr/issues/68

**Comment by danshick** [comment]
> maybe related: [emersion/xdg-desktop-portal-wlr#68](https://github.com/emersion/xdg-desktop-portal-wlr/issues/68)

I don't think so. We're not seeing any segfault in xdp. Seems instead pipewire is crashing, and xdpw exits if it cannot connect to pipewire.

**Comment by amxj9** [comment]
I've got a .pcap file from bustle, but not sure how to upload that type of file here?

In any case, I'm now using `exec sway` on login, and it works without issue - xdpw for screencasting - so happy if you want to close otherwise.  And in fact I have disabled xdg-desktop-portal-gtk as it appeared to be causing some other issues (printing issue on firefox).  No doubt I could make life easier if I used a display manager :-)
